### PR TITLE
mzML fix for special chars, Advanced import with MsProcessor to filter scans and apply processing during import

### PR DESCRIPTION
**Pull request by robinschmid**
This PR is together with olena

- fix mzml special chars 
- mzml now reads data directly and applies processing + memory mapping after each scan is completed (if scan is not filtered out by scan filter!)  
- add MsProcessor to harmonize processing of spectra during import: sort, crop, mass detect ...
- Add advanced import to imzML
- clean up mzXML import and add scan filters
- add advanced import tests
- FactorOfLowestMassDetector now works on profile mode data, uses ExactMassDetector noise 0 first and then itself
- ThermoRawFileImport and ZipImportTask now fully use MSDKMzMLImportTask

### CODE REVIEW

**Comment by robinschmid** [comment]
Finalized from my side

**Comment by robinschmid** [comment]
From the older PR, by Steffen

IMS data import does currently not work:

```java
2024-01-22 20:09:47 WARNING io.github.mzmine.modules.io.import_rawdata_mzml.MSDKmzMLImportTask importStreamOrFile
Error during mzML read: Use double buffers directly instead java.lang.UnsupportedOperationException: Use double buffers directly instead at 
io.github.mzmine.modules.io.import_rawdata_mzml.msdk.data.BuildingMzMLMsScan.getMzValues(BuildingMzMLMsScan.java:217) 
at io.github.mzmine.modules.io.import_rawdata_mzml.ConversionUtils.mzmlScanToMobilityScan(ConversionUtils.java:159) at 
io.github.mzmine.modules.io.import_rawdata_mzml.MSDKmzMLImportTask.buildIonMobilityFile(MSDKmzMLImportTask.java:350) 
at io.github.mzmine.modules.io.import_rawdata_mzml.MSDKmzMLImportTask.importStreamOrFile(MSDKmzMLImportTask.java:191) 
at io.github.mzmine.modules.io.import_rawdata_mzml.MSDKmzMLImportTask.run(MSDKmzMLImportTask.java:128) at 
io.github.mzmine.taskcontrol.impl.WorkerThread.run(WorkerThread.java:64)
```

**Comment by robinschmid** [comment]
Check out ImportSpeedTestMain for speed tests on your system. 
You can define the test files as a list of resource files or absolute pahts and then call the test in a loop testImportSpeed

File: src/main/java/io/github/mzmine/modules/io/import_rawdata_mzxml/SimpleBuildingScan.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import io.github.mzmine.datamodel.msms.MsMsInfo;
+import java.util.logging.Logger;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+
+/**
+ * Only used to build scans and contain metadata, e.g., during loading of scans
+ */
+public class SimpleBuildingScan extends MetadataOnlyScan {
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/modules/io/import_rawdata_mzxml/SimpleBuildingScan.java:44]: saw this also in the imzml import. maybe put it in a more general package

File: src/main/java/io/github/mzmine/project/impl/RawDataFileImpl.java [hunk 2]
```diff
@@ -198,18 +197,19 @@ public synchronized void addScan(Scan newScan) throws IOException {
 
     // check for zero intensity because this might indicate incorrect conversion by msconvert
     // when not using peak picking as the first step
-    if (!containsZeroIntensity) {
-      double[] intensities = newScan.getIntensityValues(new double[0]);
-      for (double v : intensities) {
-        if (v <= 0) {
+    // only check for centroid data as this might be a conversion error
+    if (!containsZeroIntensity && spectraType.isCentroided()) {
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/project/impl/RawDataFileImpl.java:201]: only check if mslevel = 1

File: src/main/java/io/github/mzmine/modules/io/import_rawdata_all/AllSpectralDataImportModule.java [hunk 3]
```diff
@@ -143,11 +243,10 @@ public ExitCode runModule(final @NotNull MZmineProject project, @NotNull Paramet
...
 
-    boolean useAdvancedOptions = parameters.getParameter(
-        AllSpectralDataImportParameters.advancedImport).getValue();
-    AdvancedSpectraImportParameters advancedParam =
-        useAdvancedOptions ? parameters.getParameter(AllSpectralDataImportParameters.advancedImport)
-            .getEmbeddedParameters() : null;
+    AdvancedSpectraImportParameters advancedParam = parameters.getEmbeddedParametersIfSelectedOrElse(
+        AllSpectralDataImportParameters.advancedImport, null);
+
+    ScanImportProcessorConfig scanProcessorConfig = createSpectralProcessors(advancedParam);
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/modules/io/import_rawdata_all/AllSpectralDataImportModule.java:249]: ah ok, so no extra load/save for the processor needed?

**Review by SteffenHeu — APPROVED** [review]
got some comments

File: src/main/java/io/github/mzmine/modules/dataprocessing/featdet_massdetection/factor_of_lowest/FactorOfLowestMassDetector.java [hunk 4]
```diff
@@ -50,6 +53,12 @@ public FactorOfLowestMassDetector(double noiseFactor) {
   }
 
   public static double[][] getMassValues(MassSpectrum spectrum, double noiseFactor) {
+    // need to apply centroiding to profile data first
+    if (spectrum.getSpectrumType() == MassSpectrumType.PROFILE) {
+      double[][] centroided = ExactMassDetector.getMassValues(spectrum, 0);
```
[the remark this answers is not in the document]
Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/dataprocessing/featdet_massdetection/factor_of_lowest/FactorOfLowestMassDetector.java:58]: True.

File: src/main/java/io/github/mzmine/modules/io/deprecated_jmzml/MzMLImportParameters.java [hunk 5]
```diff
@@ -25,15 +25,16 @@
...
 import io.github.mzmine.parameters.Parameter;
 import io.github.mzmine.parameters.impl.SimpleParameterSet;
 import io.github.mzmine.parameters.parametertypes.filenames.FileNamesParameter;
 import io.github.mzmine.util.ExitCode;
+import java.io.File;
+import java.util.List;
 import javafx.stage.FileChooser;
 import javafx.stage.FileChooser.ExtensionFilter;
 
+@Deprecated
```
[the remark this answers is not in the document]
Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/io/deprecated_jmzml/MzMLImportParameters.java:37]: Can keep it as reference or remove. does not matter to me

File: src/main/java/io/github/mzmine/modules/io/import_rawdata_all/AllSpectralDataImportModule.java [hunk 6]
```diff
@@ -114,6 +124,96 @@ public Class<? extends ParameterSet> getParameterSetClass() {
   }
 
 
+  /**
+   * Define filters and processors for scans
+   */
+  public static @NotNull ScanImportProcessorConfig createSpectralProcessors(
+      @Nullable final ParameterSet advanced) {
```
[the remark this answers is not in the document]
Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/io/import_rawdata_all/AllSpectralDataImportModule.java:131]: just changed it completely to AdvancedSpectraImportParameters as parameter

File: src/main/java/io/github/mzmine/modules/io/import_rawdata_bruker_tdf/TDFUtils.java [hunk 7]
```diff
@@ -387,11 +373,10 @@ public List<BuildingMobilityScan> loadSpectraForTIMSFrame(final long frameId,
...
 
     for (int i = 0; i < dataPoints.size(); i++) {
-      if (msDetector != null) {
-        spectra.add(new BuildingMobilityScan(i,
-            msDetector.getMassValues(dataPoints.get(i)[0], dataPoints.get(i)[1])));
-      } else {
-        spectra.add(new BuildingMobilityScan(i, dataPoints.get(i)[0], dataPoints.get(i)[1]));
+      SimpleSpectralArrays data = dataPoints.get(i);
+      if (processorConfig.hasProcessors()) {
+        data = processorConfig.processor().processScan(frame, data);
```
[the remark this answers is not in the document]
Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/io/import_rawdata_bruker_tdf/TDFUtils.java:378]: The config contains the processors and scan filter. Usually we want to apply filtering first and then processing. Not sure why the processors should contain the config with filters?

image

File: src/main/java/io/github/mzmine/modules/io/import_rawdata_mzxml/SimpleBuildingScan.java (hunk 1 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/io/import_rawdata_mzxml/SimpleBuildingScan.java:44]: ah ok its also used in imzML I can push it to the datastructures

File: src/main/java/io/github/mzmine/project/impl/RawDataFileImpl.java (hunk 2 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/project/impl/RawDataFileImpl.java:201]: done

**Pull request merged**